src/dataflow/src/source/s3.rs

```diff
@@ -752,82 +767,41 @@ async fn download_object(
     };
 
     if let Some(body) = obj.body {
-        let mut reader = body.into_async_read();
-        // unwrap is safe because content length is not allowed to be negative
-        let mut buf = Vec::with_capacity(obj.content_length.unwrap_or(1024).try_into().unwrap());
-
-        match reader.read_to_end(&mut buf).await {
-            Ok(_) => {
-                if buf.is_empty() {
-                    return (DownloadStatus::Ok, None);
-                }
-
-                let bytes_read = buf.len() as u64;
+        // sending EOF after empty objects causes empty records in dataflow
```

Comment:
This is a general problem that we shouldn't handle in the S3 specific code. I opened a PR a couple of days ago that changes this behaviour https://github.com/MaterializeInc/materialize/pull/8142

Reply:
I agree, I've gone and approved that one.